### PR TITLE
 Introduced XACRO parsing to ROS2 Gem

### PR DESCRIPTION
**Pull request by michalpelka**
The URDF importer is able to call "xacro" to convert xacro to urdf. It will show the page in the wizard with parameters.

### CODE REVIEW

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.cpp:45]: nit: we could name `i` to `rowIndex` instead just to improve readability slightly

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.cpp:54]: nit: We have `GetParams` and `SetParameters`, it'd be good to be consistent with the abbreviation or not, so maybe they're both `GetParameters` and `SetParameters`?

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.cpp:69]: question: Just double checking we want to leave the prints in permanently. If they're helpful to see in the log that's great but just wanted to check if they should be removed after debugging for example.

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h:25]: suggestion: An API comment would be really helpful here just to state briefly what XacroParamsPage does and is responsible for

File: Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp:108]: update: Looks like this print should be removed?

File: Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp:86]: suggestion: For now this is fine but if in future there are more urdf formats we might want to consider refactoring this so it is easy to add new cases without modifying an if/else or switch block (if there's just there's two it's definitely not worth it but just something that could help maintainability in future potentially

File: Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp:234]: nit: Should this printf now be removed?

File: Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp:86]: suggestion: We could use the helper `IsFileXacro` call here instead of having the same logic again (`if (IsFileXacro(m_urdfPath)`)

File: Gems/ROS2/Code/Source/RobotImporter/Utils/SourceAssetsStorage.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/Utils/SourceAssetsStorage.cpp:77]: question: This code looks very similar to that from another PR? Maybe once the other is merged this diff will go away? Just looks like this PR has some other unrelated changes in it potentially?

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:121]: nit: Would be nice to pass the filename by `const` ref (or use a `string_view` potentially)

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:138]: nit: Here it might be clearer to use `'\0'` to indicate EOF character

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.h
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.h:23]: uber nit: This types of comments in .h files tend to start with upper case characters (follow normal sentence rules) - so this just becomes `//! Parsed URDF from successful xacro's output.`

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.h
Comment (hultonha) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.h:35]: suggestion: I believe it's preferred to use `explicit` when using operator `bool` to help guard against any weird implicit conversions (this article has a not rundown - https://www.kdab.com/explicit-operator-bool/)

File: Gems/ROS2/Code/Tests/UrdfParserTest.cpp
Comment (hultonha) [inline Gems/ROS2/Code/Tests/UrdfParserTest.cpp:450]: Really awesome to see the tests! If there's time to check any other cases (maybe some incorrect data as input) to see how the code handles it, that would be great, but it's always a balance with how far you want to go

**Review by hultonha — COMMENTED** [review]
Overall looks really solid, I just have a few questions/suggestions and wanted to check about some code that maybe shouldn't be in this PR. Happy to approve once you've had a look through my comments. Thanks!

File: Gems/ROS2/Code/Tests/UrdfParserTest.cpp
Comment (moraaar) [inline Gems/ROS2/Code/Tests/UrdfParserTest.cpp:452]: `st` -> `xacroParams`

File: Gems/ROS2/Code/Tests/UrdfParserTest.cpp
Comment (moraaar) [inline Gems/ROS2/Code/Tests/UrdfParserTest.cpp:456]: `p` -> `params`

**Review by moraaar — COMMENTED** [review]
A few more minor suggestions, agree with Tom's comments. Great job!

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h
Comment (lumberyard-employee-dm) [inline Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h:39]:
```suggestion
        QTableWidget* m_table{};
```

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.h
Comment (lumberyard-employee-dm) [reply Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.h:35]: It is definitely perfected to use explicit here and something we should add to the [C++ best practices guide](https://github.com/o3de/sig-core/blob/main/governance/C%2B%2B-Best-Practices-Guide.md).

Also this function should be marked **const** as well.
```suggestion
        explicit operator bool() const
```

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (lumberyard-employee-dm) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:89]: It would be better to accept an `AZStd::span<char>` here and that would allow the any type of contiguous structure of `char` to be supplied to this method(`AZStd::vector<char>`, `AZStd::array<char, N>`, `AZStd::string`, `AZStd::fixed_string<N>`, `AZStd::string_view`, `AZStd::fixed_vector<char, N>`, etc...)

Here is the API for https://en.cppreference.com/w/cpp/container/span

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (lumberyard-employee-dm) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:66]: I recommend renaming the `cerr` in case some previous include file includes `<iostream>` and performs a `using namespace std;`, which would cause the `std::cerr` variable to be brought into scope.

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h
Comment (spham-amzn) [reply Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h:31]: as @hultonha pointed out, just the signature is specified, but there is no body. The compiler will report this as any error, so if its not used or implemented, it should be removed.

File: Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp
Comment (spham-amzn) [inline Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp:100]: nit: Should add an additional '\n'
```suggestion
                    report += QString::fromLocal8Bit(outcome.m_logErrorOutput.data(), static_cast<int>(outcome.m_logErrorOutput.size())) + "\n";
```

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (lumberyard-employee-dm) [reply Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:138]: Or avoid the issue entirely and change the charBuffer type to an AZStd::string.
The `+1` on the length wouldn't be needed as the AZStd::string will set `charBuffer[length] = char{}` in  the [resize_no_construct](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/string/string.h#L1037-L1060) function

```
        AZStd::string charBuffer;
        charBuffer.resize_no_construct(length);
        fileStream.Read(length, charBuffer.data());
```

Or better yet, use the [resize_and_overwrite](https://en.cppreference.com/w/cpp/string/basic_string/resize_and_overwrite) function which will avoid the string from ever being in an invalid state

```suggestion
        AZStd::string charBuffer;
        auto ReadFile = [&fileStream](char* buffer, size_t size)
        {
            return systemFileStream.Read(size, buffer);
        };
        charBuffer.resize_and_overwrite(length, ReadFile);
```

File: Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp
Comment (spham-amzn) [inline Gems/ROS2/Code/Source/RobotImporter/RobotImporterWidget.cpp:361]: Linux is case sensitive, suggest capitalizing the filename and then checking against an all caps ".XACRO"

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.cpp
Comment (lumberyard-employee-dm) [reply Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.cpp:65]: > `name` and `value` can be const.

Also this needs to call `.constData` on the converted QByteArray returned from the `toUtf8` call.
This will cause an an implicit conversion warning on clang which will result in a compile error
```suggestion
            AZStd::string name(p1->text().toUtf8().constData());
            AZStd::string value(p2->text().toUtf8().constData());
```

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (lumberyard-employee-dm) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:64]: These printfs are missing newlines to make sure the next output is on a separate line.

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (lumberyard-employee-dm) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:30]: repeat: these printfs will all be on the same line.
A newline needs to be added.
```suggestion
        AZ_Printf("ParseXacro", "xacro executable : %s\n", xacroPath.c_str());
        AZ_Printf("ParseXacro", "Convert xacro file : %s\n", filename.c_str());
```

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (lumberyard-employee-dm) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:52]: A newline is needed here.
```suggestion
        AZ_Printf("ParseXacro", "calling file : %s\n", outcome.m_called.c_str());
```

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (spham-amzn) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:85]: ```which``` is Linux specific. Even though this is Linux supported only, it still needs to gate this for just Linux. The suggested approach is to apply the PAL pattern for this behavior if possible, and asserting or erroring out on non-supported platforms until it becomes supported

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (spham-amzn) [reply Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:93]: Also could be constexpr as well

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (lumberyard-employee-dm) [reply Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:85]: I have a suggesstion of making the path to the `xacro` executable a setting within the Settings Registry

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (lumberyard-employee-dm) [inline Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp:13]: There is actually no need to use Qt here.

The AzFramework library has a [ProcessWatcher](https://github.com/o3de/o3de/blob/development/Code/Framework/AzFramework/AzFramework/Process/ProcessWatcher.h#L69-L70) class that can launch processes across all the support platforms

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.cpp
Comment (michalpelka) [reply Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.cpp:34]: It is overridden from QWizardPage::isComplete, can't do.

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h
Comment (michalpelka) [reply Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h:31]: Sorry about that!

File: Gems/ROS2/Code/Source/RobotImporter/Utils/SourceAssetsStorage.cpp
Comment (michalpelka) [reply Gems/ROS2/Code/Source/RobotImporter/Utils/SourceAssetsStorage.cpp:77]: Yes, it was rebased on top of PR that we already merged and squashed. I will rebase.

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (michalpelka) [reply Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp]: Unfortunately, rapid json needs a zero-terminated, mutable char array. I moved copy inside the parsing function.

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp
Comment (michalpelka) [reply Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.cpp]: Implemented @lumberyard-employee-dm suggestion. Thanks.

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h
Comment (michalpelka) [reply Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h]: Thanks, fixed.

File: Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h
Comment (michalpelka) [reply Gems/ROS2/Code/Source/RobotImporter/Pages/XacroParamsPage.h:27]: Added line of documentation.

File: Gems/ROS2/Code/Source/RobotImporter/Utils/SourceAssetsStorage.cpp
Comment (michalpelka) [reply Gems/ROS2/Code/Source/RobotImporter/Utils/SourceAssetsStorage.cpp]: Code was removed.

File: Gems/ROS2/Code/Source/RobotImporter/Utils/SourceAssetsStorage.cpp
Comment (hultonha) [reply Gems/ROS2/Code/Source/RobotImporter/Utils/SourceAssetsStorage.cpp]: Okay great thanks :+1:

File: Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.h
Comment (hultonha) [reply Gems/ROS2/Code/Source/RobotImporter/xacro/XacroUtils.h]: Could you make a PR for this update to the C++ best practices/coding standards @lumberyard-employee-dm? 🙂

**Review by hultonha — APPROVED** [review]

**Review by moraaar — APPROVED** [review]

**Review by spham-amzn — APPROVED** [review]

**Review by lemonade-dm — APPROVED** [review]

**Comment by spham-amzn** [comment]
@michalpelka I was testing this PR locally and noticed that I had to install an additional ROS package:
```
sudo apt install ros-$ROS_DISTRO-xacro
```
Can we add that to the README.md for this Gem as part of this PR?

**Comment by michalpelka** [comment]
@spham-amzn updated the readme with the package you pointed out. Thanks!

**Comment by hultonha** [comment]
Just started another build 👍

**Comment by adamdbrw** [comment]
Seems that all is in order here, could this be merged?

**Comment by hultonha** [comment]
> Seems that all is in order here, could this be merged?

Merged 👍

**Pull request merged**
